### PR TITLE
Change Version string to json for client and daemon #1272

### PR DESCRIPTION
**Pull request by stevenceuppens**
Client prints client and daemon version to the prompt in JSON format

Fixes: #1272

Signed-off-by: Steven Ceuppens <steven.ceuppens@icloud.com>

```release-note
Update version cmd output to json
```

### CODE REVIEW

File: cilium/cmd/version.go
Comment (manalibhutiyani) [inline cilium/cmd/version.go]: Hi Steven, Thanks for helping out! Could you put this line back.
We generally follow this convention

standard packages

cilium packages

third party packages

File: daemon/debuginfo.go
Comment (manalibhutiyani) [inline daemon/debuginfo.go]: Why was this format changed?

**Comment by stevenceuppens** [comment]
@manalibhutiyani @aanm 

This is a first draft, please provide feedback!

**Review by manalibhutiyani — COMMENTED** [review]
Changes look fine overall to me.  You will also need to rebase this with master once everything looks ok.

File: daemon/debuginfo.go
Comment (stevenceuppens) [reply daemon/debuginfo.go]: My current output looks like this

```
$ cilium version
Client: {"Version":"0.13.90","Revision":"377eaf9","GoRuntimeVersion":"go1.9","Arch":"linux/amd64"}
Daemon: {"Version":"0.13.90","Revision":"377eaf9","GoRuntimeVersion":"go1.9","Arch":"linux/amd64"}
```
without the change the server sends back braces arround the Revision 
```
$ cilium version
Client: {"Version":"0.13.90","Revision":"377eaf9","GoRuntimeVersion":"go1.9","Arch":"linux/amd64"}
Daemon: {"Version":"0.13.90","Revision":"(377eaf9)","GoRuntimeVersion":"go1.9","Arch":"linux/amd64"}
```

Comment (manalibhutiyani) [reply daemon/debuginfo.go]: Ah... ok.. thanks!

Comment (manalibhutiyani) [reply daemon/debuginfo.go]: I think this is breaking the jenkins tests. It could be because `dr.CiliumVersion ` is being used at other places, and the format is assumed in a specific way.
  Checking right now.

**Comment by raybejjani** [comment]
Thanks for the PR, @stevenceuppens.
Would it be possible to expand this PR somewhat and support a `--json` flag? This flag would emit a single JSON object (with client and server as top level fields) so that we can parse the version info easily with jq or other tools. The non `--json` output would remain "human readable".
What I have in mind:
```
$ cilium version --json
{
  "client": {
    "Version": "0.13.90",
    "Revision": "377eaf9",
    "GoRuntimeVersion": "go1.9",
    "Arch": "linux/amd64"
  },
  "daemon": {
    "Version": "0.13.90",
    "Revision": "377eaf9",
    "GoRuntimeVersion": "go1.9",
    "Arch": "linux/amd64"
  }
}
```

There is a global --json that's hidden, but you can also simply include it in the version command (see https://github.com/cilium/cilium/blob/master/cilium/cmd/endpoint_healthz.go#L42).


@joestringer I think you had a more long-term scheme for how to support the global --json flag, does what I said above work with that?

**Comment by stevenceuppens** [comment]
@raybejjani @manalibhutiyani 

I'm thinking about adding a version endpoint to the daemon api, that returns a version string in the same format as the client version string... this way the output can be made more consistent, and we don't need to fetch the full debugInfo body to get the version info.

Would it also make sense to add a DateTime field to the CiliumVersion struct (and json)? 

current output:
```
$ cilium version 
Client: 0.13.90 ca8aa8c Fri, 5 Jan 2018 01:11:06 +0100 go version go1.9 linux/amd64
Daemon: 0.13.90 ca8aa8c linux/amd64
```
```
$ cilium version --json
{"Client":{"Version":"0.13.90","Revision":"ca8aa8c","GoRuntimeVersion":"go1.9","Arch":"linux/amd64","BuildDate":""},"Daemon":{"Version":"0.13.90","Revision":"ca8aa8c","GoRuntimeVersion":"go1.9","Arch":"linux/amd64","BuildDate":""}}
```

**Comment by eloycoto** [comment]
Hi @stevenceuppens 

It'll nice if you can add the output as other commands are made: 

JSON flag: 
https://github.com/cilium/cilium/blob/master/cilium/cmd/service_list.go#L40

And printer: 

https://github.com/cilium/cilium/blob/master/cilium/cmd/service_list.go#L97-L102

Many thanks!

**Comment by manalibhutiyani** [comment]
@stevenceuppens `Would it also make sense to add a DateTime field to the CiliumVersion struct (and json)?`
That sounds good!

**Comment by ianvernon** [comment]
Sorry for the noise from ciliumbot, I am working on some GitHub / Jenkins configuration. This is due to GitHub Pull Request Builder integration. See https://github.com/janinko/ghprb for more information.

**Comment by raybejjani** [comment]
@stevenceuppens 
Adding version to the daemon API seems like a good idea. @cilium/api would have to review it, though, and they might disagree. Note that you'll have to regenerate the API code and it is finnicky. You need go-swagger 0.12.0 and when I've done this I've had to do the clone/checkout into my gopath manually. We list the version dependency in http://cilium.readthedocs.io/en/latest/contributing/#building-documentation but the command isn't likely to grab the correct one. 

Is the DateTime field you're proposing the build time? That would be good to have but it might be confusing to read in the current version output since we already have a commit time. This is simple to fix, obviously, so we shouldn't block on it or anything.

The tests fail with
`13:52:40 [K8s multi node Tests] error: error validating "/home/vagrant/go/src/github.com/cilium/cilium/tests/k8s/tests/deployments/gsg/demo.yaml": error validating data: found invalid field nodeSelector for v1beta1.DeploymentSpec; if you choose to ignore these errors, turn validation off with --validate=false`
which isn't related to this PR at all, I don't think. I suspect a rebase to master will sort it out.

**Comment by stevenceuppens** [comment]
Current output, if Daemon is running

```
$ cilium version 
Client: 0.13.90 9628c81 Tue, 9 Jan 2018 16:09:42 +0100 go version go1.9 linux/amd64
Daemon: 0.13.90 9628c81 Tue, 9 Jan 2018 16:09:42 +0100 go version go1.9 linux/amd64
```
```
$ cilium version -o json
{
  "Client": {
    "Version": "0.13.90",
    "Revision": "9628c81",
    "GoRuntimeVersion": "go1.9",
    "Arch": "linux/amd64",
    "BuildDate": "Tue, 9 Jan 2018 16:09:42 +0100"
  },
  "Daemon": {
    "Version": "0.13.90",
    "Revision": "9628c81",
    "GoRuntimeVersion": "go1.9",
    "Arch": "linux/amd64",
    "BuildDate": "Tue, 9 Jan 2018 16:09:42 +0100"
  }
}
```

Output, If Daemon is not running:
```
$ cilium version 
Client: 0.13.90 9628c81 Tue, 9 Jan 2018 16:09:42 +0100 go version go1.9 linux/amd64
Daemon: Not responding
```
```
$ cilium version -o json
{
  "Client": {
    "Version": "0.13.90",
    "Revision": "9628c81",
    "GoRuntimeVersion": "go1.9",
    "Arch": "linux/amd64",
    "BuildDate": "Tue, 9 Jan 2018 16:09:42 +0100"
  },
  "Daemon": {
    "Version": "",
    "Revision": "",
    "GoRuntimeVersion": "",
    "Arch": "",
    "BuildDate": ""
  }
}
```

File: cilium/cmd/version.go
Comment (rlenglet) [inline cilium/cmd/version.go]: Make this a constant (since it's also used in getDaemonVersionAsStruct() below) or return `nil` as your magic value.

File: pkg/version/version.go
Comment (rlenglet) [inline pkg/version/version.go]:
```
Version: strings.Join(results[1:4], "."),
```
or:
```
Version: fmt.Sprintf("%s.%s.%s", results[1], results[2], results[3]),
```

File: pkg/version/version.go
Comment (rlenglet) [inline pkg/version/version.go]: If parsing the build date+time is so tricky, I'd refactor it to use the ISO 8601 / RFC 3339 format (with "T" as separator).

File: pkg/version/version.go
Comment (rlenglet) [inline pkg/version/version.go]: Specify the date format used here. I'd recommend using a standard format, specifically RFC 3339 with "T" as separator. That will be unambiguous and trivial to parse.

To change the date/time format to RFC 3339, you'll need to change `Makefile.defs` to call `git show` as:
```
git show -s --format='format:%h %aI'
```
instead of:
```
git show -s --format='format:%h %aD'
```

**Review by rlenglet — APPROVED** [review]

**Review by aalemayhu — CHANGES_REQUESTED** [review]
Please address the remaining feedback and rebase against the master branch. Would also be good if you can drop the merge commits.

**Review by rlenglet — APPROVED** [review]
Very nice. Thanks!

**Comment by raybejjani** [comment]
test-me-please

File: pkg/version/version.go
Comment (raybejjani) [inline pkg/version/version.go]: From what I can tell, this is the git Author date, and not the build date. Did you intend for it to be the git date? If so, could we change the name to better reflect that?
If not, we'll have to inject the output of `date -u '+%Y-%m-%dT%H:%M:%S%z'` into the `Build` variable in Makefiles.def. We'd also have to parse it appropriately. To simplify parsing, it might easier to include a simple delimiter in the raw string (tabs might work, since they won't ruin the human readable version). We can then trivially split in this file, instead of constructing a more complex regex.

**Review by raybejjani — CHANGES_REQUESTED** [review]
One issue with the BuildDate variable, otherwise this is fantastic. Thank you!

File: pkg/version/version.go
Comment (stevenceuppens) [reply pkg/version/version.go]: @raybejjani  I changed the name for this one to AuthorDate. We can always add build date if you see some use cases.

Comment (raybejjani) [reply pkg/version/version.go]: Excellent. Thank you. I've historically found build dates useful when the version info wasn't good enough (or when I worked on something that released multiple builds of the same thing without marking them). That's not the case here, I don't think.

**Review by raybejjani — APPROVED** [review]
Oh, also, could you squash these changes once everyone has approved? I think a single commit is fine, but it's up to you.

**Comment by raybejjani** [comment]
test-me-please

**Comment by stevenceuppens** [comment]
I made on last change in DebugInfo (no need to parse and rebuild version string)

https://github.com/stevenceuppens/cilium/blob/34529cf442cd48298c2c63799592b2d020b200d5/daemon/debuginfo.go#L44

**Review by ianvernon — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

**Review by aalemayhu — APPROVED** [review]

File: pkg/version/version.go
Comment (aanm) [inline pkg/version/version.go]: Can you check for `len(fields)` to avoid panics?

File: pkg/version/version.go
Comment (aanm) [inline pkg/version/version.go:41]: There's no need for this function to be public, right?

**Review by aanm — CHANGES_REQUESTED** [review]
This looks amazing, thank you for take care of this!

**Comment by tgraf** [comment]
Change looks good to merge. I've triggered a manual Jenkins replay as Jenkins seems to have gotten stuck on this PR.

File: pkg/version/version.go
Comment (stevenceuppens) [reply pkg/version/version.go:41]: it is used [here](https://github.com/stevenceuppens/cilium/blob/34529cf442cd48298c2c63799592b2d020b200d5/cilium/cmd/version.go#L82), outside of this package

Comment (aanm) [reply pkg/version/version.go:41]: My bad, the files have the same name so that's why I miss it.

File: pkg/version/version.go
Comment (stevenceuppens) [reply pkg/version/version.go]: Fixed!

https://github.com/stevenceuppens/cilium/blob/ee1c0b6cca07850ed56e735330aa0764446f52bc/pkg/version/version.go#L44

**Comment by raybejjani** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by raybejjani** [comment]
@stevenceuppens thank you for all this work :)

**Pull request merged**
